composer/utils/file_helpers.py

```diff
@@ -0,0 +1,305 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+"""Utility for retrieving files."""
+
+import os
+import pathlib
+from typing import Iterator, Optional, Union
+
+import requests
+import tqdm
+
+from composer.core.time import Timestamp
+from composer.utils import dist
+from composer.utils.iter_helpers import iterate_with_pbar
+from composer.utils.object_store import ObjectStore
+
+__all__ = [
+    'GetFileNotFoundException',
+    'get_file',
+    'ensure_folder_is_empty',
+    'format_name_with_dist',
+    'format_name_with_dist_and_time',
+    'is_tar',
+]
+
+
+class GetFileNotFoundException(RuntimeError):
+    """Exception if :meth:`get_file` failed due to a not found error."""
+    pass
+
+
+def is_tar(name: Union[str, pathlib.Path]) -> bool:
+    """Returns whether ``name`` has a tar-like extension.
+
+    Args:
+        name (str | pathlib.Path): The name to check.
+
+    Returns:
+        bool: Whether ``name`` is a tarball.
+    """
+    return any(str(name).endswith(x) for x in (".tar", ".tgz", ".tar.gz", ".tar.bz2", ".tar.lzma"))
+
+
+def ensure_folder_is_empty(folder_name: Union[str, pathlib.Path]):
+    """Ensure that ``folder_name`` is empty.
+
+    Hidden files and folders (those beginning with ``.``) and ignored. Sub-folders are checked recursively.
+
+    Args:
+        folder_name (str | pathlib.Path): The folder to create or to ensure that it is empty.
+
+    Raises:
+        FileExistsError: If ``folder_name`` contains any non-hidden files, recursively.
+    """
+    for root, dirs, files in os.walk(folder_name, topdown=True):
+        # Filter out hidden folders
+        dirs[:] = (x for x in dirs if not x.startswith('.'))
+        for file in files:
+            if not file.startswith("."):
+                raise FileExistsError(f"{folder_name} is not empty; {os.path.join(root, file)} exists.")
```

Comment:
This function doesn't look recursive.

Reply:
`os.walk` is recursive

Reply:
Ah, I see. Well in that case I'm pretty sure this function wouldn't behave properly for non-hidden files belonging to hidden directories. The `dirs[:] = ...` line doesn't actually do anything.

Reply:
It should -- see https://docs.python.org/3/library/os.html#os.walk.

> When topdown is True, the caller can modify the dirnames list in-place (perhaps using [del](https://docs.python.org/3/reference/simple_stmts.html#del) or slice assignment), and [walk()](https://docs.python.org/3/library/os.html#os.walk) will only recurse into the subdirectories whose names remain in dirnames; this can be used to prune the search, impose a specific order of visiting, or even to inform [walk()](https://docs.python.org/3/library/os.html#os.walk) about directories the caller creates or renames before it resumes [walk()](https://docs.python.org/3/library/os.html#os.walk) again. Modifying dirnames when topdown is False has no effect on the behavior of the walk, because in bottom-up mode the directories in dirnames are generated before dirpath itself is generated.

Reply:
Oh that's so jank but yeah that works.